### ISSUE TITLE
exception when attempting to train model

### ISSUE THREAD

**Issue by ghosttie** [MicrosoftDocs/intellicode#252]
This is in Visual Studio - I tried using the "Report a Problem..." feature but it doesn't work ("Oops, we have lost the connection to Visual Studio." every time)

It's happened for multiple solutions (all the ones I've tried so far)

```
[TrainCommand I] Beginning extraction for csharp/intellisense-members.
[AutomaticModelTrainingController W] Error while training: System.ArgumentException: Could not find extraction provider for csharp/intellisense-members
   at Microsoft.VisualStudio.IntelliCode.Extractor.TrainCommand.<ExtractAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.IntelliCode.AutomaticModelTraining.AutomaticModelTrainingController.<>c__DisplayClass68_0.<<CreateExtractionInternalAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__68.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask`1.<JoinAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.VisualStudio.IntelliCode.AutomaticModelTraining.AutomaticModelTrainingController.<CreateExtractionInternalAsync>d__68.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.IntelliCode.AutomaticModelTraining.AutomaticModelTrainingController.<CreateExtractionInternalAsync>d__68.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.VisualStudio.IntelliCode.AutomaticModelTraining.AutomaticModelTrainingController.<>c__DisplayClass67_0.<<CreateExtractionAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.IntelliCode.TaskNotifications.TaskNotificationService.<AwaitTaskWithNotificationAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.ReentrantSemaphore.NotRecognizedSemaphore.<>c__DisplayClass1_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.AwaitExtensions.ExecuteContinuationSynchronouslyAwaiter.GetResult()
   at Microsoft.VisualStudio.Threading.ReentrantSemaphore.NotRecognizedSemaphore.<ExecuteAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.IntelliCode.AutomaticModelTraining.AutomaticModelTrainingController.<CreateExtractionAsync>d__67.MoveNext()
```

**Product and Version** [VS]:
**OS Version** [Windows]:
**IntelliCode Extension Version**: 2.2.958.9542
**Target Platform or Language** [C#]:

**Steps to Reproduce / Scenario:**

1. click menu View - Other Windows - IntelliCode
2. click Refresh learned patterns (although for me it now says Retry)

**Comment by catycaldwell** [comment]
@ghosttie, thank you for reporting this. We will have someone from the team investigate.

**Comment by AnqiLi6** [comment]
@ghosttie Can you try to reproduce this issue and copy this folder to us? %localappdata%\microsoft\visualstudio\<instanceID>\ComponentModelCache\

If you are not sure what the instanceID is, you can just give us all the folders under %localappdata%\microsoft\visualstudio.
Thank you very much!

**Comment by ghosttie** [comment]
[ComponentModelCache.zip](https://github.com/MicrosoftDocs/intellicode/files/5646367/ComponentModelCache.zip)

**Comment by DavidObando** [comment]
Thank you @ghosttie, from the logs it appears that Visual Studio is having issues finding one of the required IntelliCode components:
```
Microsoft.CodeAnalysis.ExternalAccess.Pythia.PythiaSignatureHelpProvider.ctor(implementation): expected exactly 1 export matching constraints:
    Contract name: Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api.IPythiaSignatureHelpProviderImplementation
    TypeIdentityName: Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api.IPythiaSignatureHelpProviderImplementation
but found 0.
   part definition Microsoft.CodeAnalysis.ExternalAccess.Pythia.PythiaSignatureHelpProvider
```
We're looking into this. We'll let you know shortly how to proceed to get yourself unblocked. Sorry for the inconvenience!

**Comment by vivlimmsft** [comment]
I'm not sure that error could be causing this issue with model training, it might be an upstream issue in the Roslyn ExternalAccess api for signature help.

It's possible that there were other composition errors that were logged in previous sessions but were overwritten. Let's clear and rebuild the cache, so we'll have a .err log containing everything.
@ghosttie, could you please open a Developer Command Prompt and run the following two commands:
```
devenv /clearcache
devenv /updateconfiguration
```
The second one might take some time to complete. After it finishes, please copy the file at `%localappdata%\microsoft\visualstudio/[instanceID]\ComponentModelCache\Microsoft.VisualStudio.Default.err` to a safe location.

After that, launch Visual Studio with `devenv /log` and try to train a model again. If that fails, please close VS then attach the following files:
* the `Microsoft.VisualStudio.Default.err` you copied above
* `%localappdata%\microsoft\visualstudio/[instanceID]\ActivityLog.xml`
* The newest file in `%temp%\vsfeedbackintellicodelogs\`

**Comment by ghosttie** [comment]
Here is the err and log file, but there wasn't an ActivityLog.xml

[logs.zip](https://github.com/MicrosoftDocs/intellicode/files/5660152/logs.zip)

**Comment by vivlimmsft** [comment]
From your log, it seems like the IntelliCode C# package isn't activating, normally there's a line in the IntelliCode log which starts with
`[2020-12-02 20:15:21.926 CSharp.Package I] [C# IntelliCode] C# IntelliCode Completions` followed by a version number.

To look further into that, I'll need a couple things: the ActivityLog.xml and install logs. 
The ActivityLog.xml is only fully written to disk when devenv.exe is started with the `/log` flag. If that's not working for you, you can also try following it with a path to write the activitylog to, such as `devenv /log %userprofile%\desktop\activitylog.xml`.

For the install logs, please run the tool at https://aka.ms/vscollect. It'll produce a zip file. It's quite large and may contain some personal data, so if you'd prefer to not upload it here, upload it to cloud storage and email the link to vivlim@microsoft.com.

Thanks for your patience in collecting these logs and the other logs you've already collected.

**Comment by ghosttie** [comment]
Thanks, I got the activity log to work using your instructions, I'll send you an email regarding the other file

[ActivityLog.zip](https://github.com/MicrosoftDocs/intellicode/files/5662241/ActivityLog.zip)

**Comment by vivlimmsft** [comment]
Thanks! We're looking at the logs now. I removed the ActivityLog.xml link from your last comment after I was able to download it.

**Comment by johndaintree** [comment]
I'm getting the same problem when trying to report a problem with Visual Studio 2019 Preview. I just start VS, (no solution loaded) and hit Help->Send Feedback->Report a Problem, and I always get: "Oops, we have lost the connection to Visual Studio."

VS details follow:

Microsoft Visual Studio Professional 2019 Preview
Version 16.9.0 Preview 2.0
VisualStudio.16.Preview/16.9.0-pre.2.0+30803.129
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-60000-00000-AA161
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.9.311.8738
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2057006+e0968ea4b5f5c523cc7efea1d91fec21e0adc648
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.9.311.8738
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.9.311.8738
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.9.311.8738
Azure Functions and Web Jobs Tools

C# Tools   3.9.0-2.20602.2+eb0a7ff1c241695e17b07b096593804fdd5cfcfd
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.31203.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.9.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62011.24130
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.21120.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.9.0-2.20602.2+eb0a7ff1c241695e17b07b096593804fdd5cfcfd
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.9.0-beta.20567.7+c02b2ef997f1dd07993e1f85b017c858bdf82530
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

**Comment by vivlimmsft** [comment]
Hi @ghosttie,

Sorry for the delay in getting back to you. We found an issue where if IntelliCode was ever installed from the Visual Studio Marketplace before upgrading to 16.8, the component which provides IntelliSense stars and model training for C# would not be installed.
A fix for this has been implemented internally and will be released in the upcoming Visual Studio 16.8.4.

@johndaintree, are you seeing the same issue where you cannot train an IntelliCode model?

**Comment by ghosttie** [comment]
Thanks, I'll look forward to it 😁

**Comment by pvoigt** [comment]
While waiting for 16.8.4, is there any solution to the issue?

**Comment by pvoigt** [comment]
> While waiting for 16.8.4, is there any solution to the issue?

And the wait is over.... the issue appears to be resolved with 16.8.4 for me at least.

**Comment by markw-t** [comment]
Thanks for confirming @pvoigt ! I'm closing out the issue.
